### PR TITLE
Add readme files describing model execution flow for ASR tasks

### PR DESCRIPTION
**Pull request by titu1994**
Signed-off-by: smajumdar <titu1994@gmail.com>

# What does this PR do ?

Adds high level architecture / model execution flow diagrams to better enable users to understand how to write model scratch training / fine-tuning steps / inference steps

**Collection**: [ASR]

# Changelog 
- Add flow diagrams for CTC/RNNT model training
- Add flow diagram for ASR inference
- Add flow diagram for Speech classification
- Add flow diagram for Speech SSL

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [x] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

### CODE REVIEW

File: examples/asr/README.md [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+The inference scripts in this directory execute in the following order. When preparing your own inference scripts, please follow this order for correct inference.
+
+```mermaid
+
+graph TD
+    A[Hydra Overrides + Config Dataclass] --> B{Config}
+    B --> |Init| C[Model]
+    B --> |Init| D[Trainer]
+    C & D --> E[Set trainer]
+    E --> |Optional| F[Change Transducer Decoding Strategy]
```
Comment (piraka9011) [inline examples/asr/README.md:18]: Maybe

```suggestion
    E --> |Optional| F[Change Decoding Strategy]
```

To keep it generalized?

Comment (piraka9011) [reply examples/asr/README.md:18]: Consider also adding other changes to this path (in order) for ex. changing the optimizer and datasets.

File: examples/asr/asr_ctc/README.md [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    A[Hydra Overrides + Yaml Config] --> B{Config}
+    B --> |Init| C[Trainer]
+    C --> D[ExpManager]
+    B --> D[ExpManager]
+    C --> E[Model]
+    B --> |Init| E[Model]
+    E --> |Constructor| F(Change Vocabulary)
+    F --> G(Setup Train + Validation + Test Data loaders)
+    G --> H(Setup Optimization)
+    H --> I[Maybe init from pretrained]
```
Comment (piraka9011) [inline examples/asr/asr_ctc/README.md:26]: Should this step be right after initializing the model? Or should it be the last thing?

Reason I ask is b/c wouldn't the `maybe_init_from_pretrained()` method override the previous behavior in nodes F/G/H?

**Comment by piraka9011** [comment]
TIL you can draw graphs in Github's MD...

I think this coupled w/ suggested change to the tutorial would be perfect!

File: examples/asr/README.md (hunk 1 above)
Comment (titu1994) [reply examples/asr/README.md:18]: CTC does not have decoding strategy change. So its proper to keep it specific to transducer. 

This is the inference pathway - the train scripts in asr_ctc and asr_transducer have their own diagrams for training/finetuning.

File: examples/asr/asr_ctc/README.md (hunk 2 above)
Comment (titu1994) [reply examples/asr/asr_ctc/README.md:26]: Maybe init from pretraind only loads the pytorch weights of the old file into the new file, it has no effect at all on any of the dataloaders / optim etc.

**Review by VahidooX — APPROVED** [review]

**Comment by titu1994** [comment]
> TIL you can draw graphs in Github's MD...

Github had a whole announcement, and I had been waiting for a use case, so I guess things worked out in the end :D 
https://github.blog/2022-02-14-include-diagrams-markdown-files-mermaid/

**Pull request merged**